server/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -615,8 +616,11 @@ private GetResult getFromTranslog(Get get, Translog.Index index, MappingLookup m
         final Engine.Searcher searcher = new Engine.Searcher("realtime_get", ElasticsearchDirectoryReader.wrap(inMemoryReader, shardId),
             config().getSimilarity(), config().getQueryCache(), config().getQueryCachingPolicy(), inMemoryReader);
         final Searcher wrappedSearcher = searcherWrapper.apply(searcher);
-        if (wrappedSearcher == searcher) {
-            searcher.close();
+        // TODO: we always wrap with field usage tracking, so we need to find another way to determine if we can safely do this
```

Comment:
@dnhatn would love to have your thoughts on this ^^

I'm wondering if we could have `TranslogLeafReader` somehow embedded in `SingleDocDirectoryReader` so that we always apply the wrapper, but then lazily load the `InMemoryLeafReader` when there is access to more than what the `TranslogLeafReader ` can offer.

Reply:
I've built a solution for this. I won't push it to this PR though as it is quite a big change to how TranslogLeafReader works. Will make a separate PR tomorrow if test show all is fine with that solution.

Reply:
I've opened #74722 for this.